Fix ansible_virtualization_role fact for DigitalOcean guests

siwyd opened the pull request:
##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
module_utils/facts.py

##### SUMMARY
DigitalOcean guests show 'host' for ansible_virtualization_role fact  at the moment. This commit fixes that. It might fix it for other cloud providers (and emulators) as well perhaps, but my use case was for DO.

DO guests expose the chassis_vendor as 'Bochs' and considering guest/kvm is already being set for role/type when the product_name is 'Bochs', I've continued using that logic.

Also, my educated guess is dat DigitalOcean uses KVM as virtualization, but I'm not 100% sure. I don't care that much about the type, for me the role was important because I want to do bare metal specific tasks if applicable and skip them otherwise.

Before fix:

```
hb@aramaki ~/git/hyperbaton-ansible (git)-[master] % ansible fedora-test -m setup -u root -a 'filter=ansible_virtualization_*'
138.68.82.12 | SUCCESS => {
    "ansible_facts": {
        "ansible_virtualization_role": "host", 
        "ansible_virtualization_type": "kvm"
    }, 
    "changed": false
}
```
After fix:
```
hb@aramaki ~/git/hyperbaton-ansible (git)-[master] % ansible fedora-test -m setup -u root -a 'filter=ansible_virtualization_*'
138.68.82.12 | SUCCESS => {
    "ansible_facts": {
        "ansible_virtualization_role": "guest", 
        "ansible_virtualization_type": "kvm"
    }, 
    "changed": false
}
```

Commit:
Fix ansible_virtualization_role fact for DigitalOcean guests

DigitalOcean guests show 'host' for ansible_virtualization_role at the
moment. This commit fixes that. It might fix it for other cloud providers
as well perhaps, but my use case was for DO.

DO guests expose the chassis_vendor as 'Bochs' and considering
guest/kvm is already being set when the product_name is 'Bochs', I've
continued using that logic.

dagwieers commented:
IIRC This was going to be reorganized, because a system can be both a guest and a host.

HyperBaton commented:
@dagwieers Ah that sounds good because the facts at the moment don't really help when dealing with nested virtualization. If that's planned for 2.3, then this pull could possibly be ignored, otherwise it would be nice to have this in for 2.3.

simonwydooghe commented:
Could this merge request be reviewed by someone?

siwyd commented:
I've rebased and reworked this. Seems like the facts have been split up in multiple files now? Could this get a review? Seems like this has missed 2.3 and 2.4 window already :(

siwyd commented:
Soooo... Any update on this?

alikins commented:
> Also, my educated guess is dat DigitalOcean uses KVM as virtualization, but I'm not 100% sure. I don't care that much about the type, for me the role was important because I want to do bare metal specific tasks if applicable and skip them otherwise.

Would like some confirmation that 'kvm' is the right guest type. 
I didn't find anything definative in a quick look around digitalocean.com.

https://en.wikipedia.org/wiki/DigitalOcean claims it is now KVM based.
https://www.quora.com/Did-Digital-Ocean-build-an-entire-cloud-orchestration-platform (from 2013)
claims it was something custom at the time.

If it is just a KVM guest, I would expect the regular KVM guest detection to work. 
That code may be buggy or the digitalocean droplets aren't quite KVM guests.
I think I see a bug in the kvm hosts detection that could be incorrectly adding the 'host' type...
(line 212 of virtual/linux.py)

But sort of depends on what is considered virt. How are you using
the  ansible_virtualization_role fact?

alikins commented on lib/ansible/module_utils/facts/virtual/linux.py:
```diff
@@ -142,6 +142,13 @@ def get_virtual_facts(self):
+
+        if chassis_vendor == 'Bochs':
+            self.facts['virtualization_type'] = 'kvm'
+            self.facts['virtualization_role'] = 'guest'
```
Would like confirmation the droplets are KVM guests.

anthonyeleven replied:
@siwyd  @alikins Such timing -- I was about to submit a PR against module_utils/facts/virtual/linux.py to check {{product_name}} to detect a DO droplet when I came across this PR.

$ cat /sys/devices/virtual/dmi/id/product_name
Droplet

siwyd commented:
@alikins https://www.digitalocean.com/help/technical/general/ confirms that DigitalOcean uses KVM. It's been a long time, but my use case was to run tasks specific to bare metal or virtual hosts.

alikins commented:
@siwyd Any way you could test the patch in https://github.com/ansible/ansible/pull/35403 to see if it fixes this?  I think this particular case is an instance of a more general bug that 35403 tries to resolve.
